### PR TITLE
Adds option to calculate in-cluster CCSD and FCI moments via Dyson

### PR DESCRIPTION
**Pull request by basilib**

### CODE REVIEW

**Comment by obackhouse** [comment]
Nice stuff - can you make sure to add `dyson` as an optional dependency in `__init__.py`, and once #97 is merged, in the `pyproject.toml`? Thanks

File: vayesta/solver/ccsd.py [hunk 1]
```diff
@@ -70,6 +73,35 @@ def kernel(self, t1=None, t2=None, l1=None, l2=None, coupled_fragments=None, t_d
...
+            # vecs_ket = expr.build_gf_vectors(nmom[0], left=False)
+            # amps_ket = [expr.eom.vector_to_amplitudes(amps[n,p], ccm.nmo, ccm.nocc) for p in range(ccm.nmo) for n in range(nmom)]
+            # self.ip_moment_amplitudes = (amps_bra, amps_ket)
+
+            # expr = CCSD["1p"](mf_clus, t1=mycc.t1, t2=mycc.t2, l1=l1, l2=l2)
+            # vecs_bra = expr.build_gf_vectors(nmom[0], left=True)
+            # amps_bra = [expr.eom.vector_to_amplitudes(amps[n,p], ccm.nmo, ccm.nocc) for p in range(ccm.nmo) for n in range(nmom)]
+            # vecs_ket = expr.build_gf_vectors(nmom[0], left=False)
+            # amps_ket = [expr.eom.vector_to_amplitudes(amps[n,p], ccm.nmo, ccm.nocc) for p in range(ccm.nmo) for n in range(nmom)]
+            # self.ea_moment_amplitudes = (amps_bra, amps_ket)
```
Comment (obackhouse) [inline vayesta/solver/ccsd.py:98]: Can this commented code be removed, or is it needed?

File: vayesta/tests/ewf/test_moments.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            cx = f.get_overlap('mo|cluster')
+            ip = np.einsum('pP,qQ,nPQ->npq', cx, cx, ip)
+            ea = np.einsum('pP,qQ,nPQ->npq', cx, cx, ea)
+
+            self.assertTrue(np.allclose(ip, fci_ip))
+            self.assertTrue(np.allclose(ea, fci_ea))
+
+if __name__ == '__main__':
+    print("Running %s" % __file__)
+    unittest.main()
```
Comment (obackhouse) [inline vayesta/tests/ewf/test_moments.py:81]: Missing newline character at the end here

**Comment by obackhouse** [comment]
FYI until #97 and #104 are merged, `ebcc` isn't being imported in the CI - so these tests are not being hit on this PR. I'll push a hacky fix for now, and we will try to get #97 and #104 merged ASAP

**Comment by obackhouse** [comment]
One of the tests is failing @basilib

File: vayesta/tests/ewf/test_moments.py (hunk 2 above)
Comment (basilib) [reply vayesta/tests/ewf/test_moments.py:81]: Python needs a newline at the end of a file? Or is this a PEP8 thing?

Comment (obackhouse) [reply vayesta/tests/ewf/test_moments.py:81]: Not sure about PEP8 specifically, but any text-based file on a UNIX platform should end with a newline character as far as I know.

**Comment by basilib** [comment]
> You can merge `master` and overwrite [1e92142](https://github.com/BoothGroup/Vayesta/commit/1e92142259f23d65793b9cf06c55de91ce5ae61b), [43aa945](https://github.com/BoothGroup/Vayesta/commit/43aa945e823d2741a5c99fb7be253e336c6778c4), [600c057](https://github.com/BoothGroup/Vayesta/commit/600c05780f1eabe69962c095491776c6ab9903a6)

Thanks! 

I haven't managed to figure out the issue with the CCSD tests. I might just need to set the equality tolerance to 1e-5 or something.

**Comment by obackhouse** [comment]
> Thanks!
> 
> I haven't managed to figure out the issue with the CCSD tests. I might just need to set the equality tolerance to 1e-5 or something.

Those errors are pretty high - should they be numerically exact?

**Comment by basilib** [comment]
> Those errors are pretty high - should they be numerically exact?

I think so, the FCI tests is almost identical and passes. They both calculate the moments via dyson with from the MF reference and with vayesta with a complete bath space for each fragment. 

Moreover, both pass on zombie.

**Comment by obackhouse** [comment]
> I think so, the FCI tests is almost identical and passes. They both calculate the moments via dyson with from the MF reference and with vayesta with a complete bath space for each fragment.
> 
> Moreover, both pass on zombie.

Yeah that's concerning then and shouldn't be looked over - I can take a look on my zombie config

**Comment by basilib** [comment]
> Yeah that's concerning then and shouldn't be looked over - I can take a look on my zombie config

Thanks, that would be useful. The test is in vayesta/tests/ewf/test_moments.py.

**Comment by obackhouse** [comment]
Oh sorry, they pass for you too but not on the CI, ignore me

File: vayesta/tests/ewf/test_moments.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    def test(self):
+
+        #RHF
+        mf = testsystems.h6_sto6g.rhf()
+
+        try:
+            from dyson.expressions import FCI
+        except ImportError:
+            vayesta.log.info("Could not import Dyson. Skipping FCI moment tests.")
+            return
```
Comment (obackhouse) [inline vayesta/tests/ewf/test_moments.py:23]: Use `pytest.skip(message)` instead

File: vayesta/tests/ewf/test_moments.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    def test(self):
+
+        #RHF
+        mf = testsystems.water_sto3g.rhf()
+
+        try:
+            from dyson.expressions import CCSD
+        except ImportError:
+            vayesta.log.info("Could not import Dyson. Skipping FCI moment tests.")
+            return
```
[the remark this answers is not in the document]
Comment (obackhouse) [reply vayesta/tests/ewf/test_moments.py:57]: By these I don't mean the decorator, seems to be troublesome with our test setup, just the following pattern:

```
try:
    import x
except ImportError:
    pytest.skip(message)
```

This means that the test suite "knows" it's been skipped - if you just return I think it's considered as passed even without any asserts.

**Pull request merged**